libass/ass_utils.h

```diff
@@ -182,19 +183,52 @@ static inline int double_to_d22(double x)
     return lrint(x * 0x400000);
 }
 
+static inline int32_t lshiftwrapi(int32_t i, int32_t shift)
+{
+    bool neg_i = i < 0;
+    if (neg_i) i = -i;
+
+    // Some compilers behave differently for negative compile-time constants
+    // (e.g. on Linux gcc, clang and tcc - but not MSYS2's gcc on Windows), see:
+    // if (shift < 0)
+    //     u = ((uint32_t)i) >> -shift;
+    // else
+    uint32_t u = ((uint32_t)i) << (((uint32_t)shift) & 31);
+    i = u > INT32_MAX ? -(int32_t)(~u) - 1 : u;
```

Comment:
Elsewhere, we just cast `uint32_t` to `int32_t`, and we seem to have had the consensus opinion that it is good enough given the definition of `int32_t` as specifically two’s complement (even in C99). If we want to be extra pedantic and avoid this implementation-defined behaviour, then those other places should be sought out and adjusted as well. Some instances are in `interpolate_alpha` in `ass_parser.c` since #261, but there may be more.

Reply:
I forgot casts from integer type to signed integer type are never UB *(unlike non-cast overflows and double→int casts)*, but "only" have an implementation defined result or raise an implementation-defined signal(?). Given fixed width types are required to be two-complements, jsut casting them seems pretty reasonable then, or if we wanted to make really sure  `*(int32_t *) &uint32_variable` would work too and be probably optimised away by the compiler.